### PR TITLE
Known bug (CMIIW)

### PR DESCRIPTION
**Pull request by rhonyabdullah**
Changing the way saving config persistent into sparse array for better performance and put activity id into outstate before super.

### CODE REVIEW

**Comment by fernandofcampos** [comment]
Shouldn't onDestroy use isFinishing() instead? 

protected void onDestroy() {
        if (isFinishing()) {
            sComponentsMap.remove(mActivityId);
        }
        super.onDestroy();
}

**Comment by ribotaccounts** [comment]
Can one of the admins verify this patch?

**Comment by rhonyabdullah** [comment]
@fernandofcampos 
`isChangingConfigurations()` in `OnDestroy()` was applied on ribot android-boilerplate and i don't wanna change anything about that, but after reading this [documentation](https://developer.android.com/reference/android/app/Activity.html) i found that `isFinishing()` get called after `OnDestroy()` rather than `isChangingConfigurations()` [link](https://developer.android.com/reference/android/app/Activity.html#isChangingConfigurations()) which is get called during the android process to check whether the activity is in the process being destroyed. So i think ribot is on the correct way at this point. CMIIW.

File: app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java [hunk 1]
```diff
@@ -23,7 +22,7 @@
 
     private static final String KEY_ACTIVITY_ID = "KEY_ACTIVITY_ID";
     private static final AtomicLong NEXT_ID = new AtomicLong(0);
-    private static final Map<Long, ConfigPersistentComponent> sComponentsMap = new HashMap<>();
+    private static final LongSparseArray<ConfigPersistentComponent> sComponentsMap = new LongSparseArray<>();
```
Comment (jemmaSlater) [inline app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: This line is too long, our guidelines state that the length of a line should not exceed 100 characters. https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md#2215-line-length-limit

File: app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java [hunk 2]
```diff
@@ -34,26 +33,23 @@ protected void onCreate(Bundle savedInstanceState) {
 
         // Create the ActivityComponent and reuses cached ConfigPersistentComponent if this is
         // being called after a configuration change.
-        mActivityId = savedInstanceState != null ?
-                savedInstanceState.getLong(KEY_ACTIVITY_ID) : NEXT_ID.getAndIncrement();
-        ConfigPersistentComponent configPersistentComponent;
-        if (!sComponentsMap.containsKey(mActivityId)) {
+        mActivityId = savedInstanceState != null ? savedInstanceState.getLong(KEY_ACTIVITY_ID) : NEXT_ID.getAndIncrement();
```
Comment (jemmaSlater) [inline app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: This line is too long, our guidelines state that the length of a line should not exceed 100 characters. https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md#2215-line-length-limit

File: app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java [hunk 3]
```diff
@@ -34,26 +33,23 @@ protected void onCreate(Bundle savedInstanceState) {
...
-            configPersistentComponent = sComponentsMap.get(mActivityId);
         }
         mActivityComponent = configPersistentComponent.activityComponent(new ActivityModule(this));
     }
 
     @Override
     protected void onSaveInstanceState(Bundle outState) {
-        super.onSaveInstanceState(outState);
         outState.putLong(KEY_ACTIVITY_ID, mActivityId);
+        super.onSaveInstanceState(outState);
```
Comment (jemmaSlater) [inline app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: I don't think this change will have any effect?

Comment (rhonyabdullah) [reply app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: Okey, check again.

File: app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java (hunk 2 above)
Comment (rhonyabdullah) [reply app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: Okey, check again.

File: app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java (hunk 1 above)
Comment (rhonyabdullah) [reply app/src/main/java/uk/co/ribot/androidboilerplate/ui/base/BaseActivity.java]: Okey, check again.

**Pull request merged**